pruning/manager.go

```diff
@@ -0,0 +1,216 @@
+	pruneHeights         []int64
+	pruneSnapshotHeights *list.List
+	mx                   sync.Mutex
+}
+
+const (
+	pruneHeightsKey         = "s/pruneheights"
+	pruneSnapshotHeightsKey = "s/pruneSnheights"
+)
+
+func NewManager(logger log.Logger) *Manager {
+	return &Manager{
+		logger:       logger,
+		opts:         types.NewPruningOptions(types.PruningNothing),
+		pruneHeights: []int64{},
+		// These are the heights that are multiples of snapshotInterval and kept for state sync snapshots.
+		// The heights are added to this list to be pruned when a snapshot is complete.
+		pruneSnapshotHeights: list.New(),
+		mx:                   sync.Mutex{},
+	}
+}
+
+// SetOptions sets the pruning strategy on the manager.
+func (m *Manager) SetOptions(opts *types.PruningOptions) {
+	m.opts = opts
+}
+
+// GetOptions fetches the pruning strategy from the manager.
+func (m *Manager) GetOptions() *types.PruningOptions {
+	return m.opts
+}
+
+// GetPruningHeights returns all heights to be pruned during the next call to Prune().
+func (m *Manager) GetPruningHeights() []int64 {
+	return m.pruneHeights
+}
+
+// ResetPruningHeights resets the heights to be pruned.
+func (m *Manager) ResetPruningHeights() {
+	m.pruneHeights = make([]int64, 0)
+}
+
+// HandleHeight determines if pruneHeight height needs to be kept for pruning at the right interval prescribed by
+// the pruning strategy. Returns true if the given height was kept to be pruned at the next call to Prune(), false otherwise
+func (m *Manager) HandleHeight(previousHeight int64) int64 {
+	if m.opts.GetPruningStrategy() == types.PruningNothing {
+		return 0
+	}
+
+	defer func() {
+		// handle persisted snapshot heights
+		m.mx.Lock()
+		defer m.mx.Unlock()
+		var next *list.Element
+		for e := m.pruneSnapshotHeights.Front(); e != nil; e = next {
+			snHeight := e.Value.(int64)
+			if snHeight < previousHeight-int64(m.opts.KeepRecent) {
+				m.pruneHeights = append(m.pruneHeights, snHeight)
+
+				// We must get next before removing to be able to continue iterating.
+				next = e.Next()
+				m.pruneSnapshotHeights.Remove(e)
+			} else {
+				next = e.Next()
+			}
+		}
+	}()
+
+	if int64(m.opts.KeepRecent) < previousHeight {
+		pruneHeight := previousHeight - int64(m.opts.KeepRecent)
+		// We consider this height to be pruned iff:
+		//
+		// - snapshotInterval is zero as that means that all heights should be pruned.
+		// - snapshotInterval % (height - KeepRecent) != 0 as that means the height is not
+		// a 'snapshot' height.
+		if m.snapshotInterval == 0 || pruneHeight%int64(m.snapshotInterval) != 0 {
+			m.pruneHeights = append(m.pruneHeights, pruneHeight)
+			return pruneHeight
+		}
+	}
+	return 0
+}
+
+func (m *Manager) HandleHeightSnapshot(height int64) {
+	if m.opts.GetPruningStrategy() == types.PruningNothing {
+		return
+	}
+	m.mx.Lock()
+	defer m.mx.Unlock()
+	m.logger.Debug("HandleHeightSnapshot", "height", height) // TODO: change log level to Debug
+	m.pruneSnapshotHeights.PushBack(height)
+}
```

Comment:
So if the node halts, is it possible that the snapshot height never gets pruned?

Reply:
Seems like we should be calling `flushPruningSnapshotHeights` here?

Reply:
These are flushed here: https://github.com/osmosis-labs/cosmos-sdk/blob/a619355eecafd7ceb987e121f0dca2a29c4c009a/store/rootmulti/store.go#L993

And loaded here: https://github.com/osmosis-labs/cosmos-sdk/blob/a619355eecafd7ceb987e121f0dca2a29c4c009a/store/rootmulti/store.go#L270

I followed the same logic to how we flush/load commit info and regular pruning heights

Reply:
idt I get it, why don't we push this metadata to disk ASAP?

Reply:
Happy to merge the PR though, and talk about this in a follow-up issue